### PR TITLE
New Rule: ExcludeFromCodeCoverage attributes should include a justification

### LINKED ISSUE #6885: New Rule: ExcludeFromCodeCoverage attributes should include a justification
Both reasoning and implementation are similar to [RSPEC-1123: "Obsolete" attributes should include explanations](https://rules.sonarsource.com/csharp/RSPEC-1123).

Implementation: https://github.com/SonarSource/sonar-dotnet/pull/6593
RSpec: https://github.com/SonarSource/rspec/pull/1624

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #6885

### CODE REVIEW

**Comment by martin-strecker-sonarsource** [comment]
Hi @Corniel, thank you for your contribution. I looked into the implementation, and the implementation is just exemplary. There is one problem, though. The [ExcludeFromCodeCoverageAttribute.Justification property](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.excludefromcodecoverageattribute.justification?view=net-7.0#applies-to) was introduced in .Net 5, while the attribute itself is available in a [much more comprehensive range of frameworks](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.excludefromcodecoverageattribute?view=net-7.0#applies-to):

| Product        | Versions                                                                |
|----------------|--------------------------------------------------------------------------|
| .NET           | Core 2.0, Core 2.1, Core 2.2, Core 3.0, Core 3.1, 5, 6, 7, 8             |
| .NET Framework | 4.0, 4.5, 4.5.1, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7, 4.7.1, 4.7.2, 4.8, 4.8.1 |
| .NET Standard  | 2.0, 2.1                                                                 |
| Xamarin.iOS    | 10.8                                                                     |
| Xamarin.Mac    | 3.0                                                                      |

While we could check for the presents of the `Justification` property on the attribute symbol, this would still generate a lot of false positives in the case of multi-target projects.

As such, we will need to deactivate the rule by default (we will just don't put the rule in the default "Sonar way" profile). Do you want to create the related RSpec or do you want me to do this for you?

**Comment by Corniel** [comment]
> As such, we will need to deactivate the rule by default (we will just don't put the rule in the default "Sonar way" profile). Do you want to create the related RSpec or do you want me to do this for you?

I've created som RSPEC's in the past, and it turned out that the process was not designed to be used by external contributes. So, if you can create it, I'm more than happy to review and contribute on that PR.

**Comment by Corniel** [comment]
@martin-strecker-sonarsource Just did a rebase.

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    protected ExcludeFromCodeCoverageAttributesNeedJustificationBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterSyntaxNodeActionInNonGenerated(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (!Language.Syntax.ArgumentExpressions(c.Node).Any()
+                    && c.SemanticModel.GetSymbolInfo(c.Node).Symbol is { } attribute
+                    && attribute.IsInType(KnownType.System_Diagnostics_CodeAnalysis_ExcludeFromCodeCoverageAttribute))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:39]: Needs a check if the `attribute` symbol has a property `Justification`. This also needs some conditional tests for `#if Net` and Net framework.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    [ExcludeFromCodeCoverage] // Noncompliant
+    void Method();
+}
+
+[ExcludeFromCodeCoverage] // Noncompliant
+struct ProgramStruct
+{
+    [ExcludeFromCodeCoverage] // Noncompliant
+    void Method() { }
+}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs:45]: Nice coverage!
Add test cases with `[ExcludeFromCodeCoverage(Justification = null)]` and `[ExcludeFromCodeCoverage(Justification = "")]` and mark them as FNs.

Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs:54]: @martin-strecker-sonarsource I've made a test that shows that it does not trigger on versions of .NET that not contain this property.

I also added support for null and whitespace justifications.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs:54]: Thanks! Thats even better.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Please make the "Justification" check language specific.

File: analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs [hunk 3]
```diff
@@ -56,10 +56,12 @@ internal sealed class VisualBasicSyntaxFacade : SyntaxFacade<SyntaxKind>
     public override IEnumerable<SyntaxNode> ArgumentExpressions(SyntaxNode node) =>
         node switch
         {
-            ObjectCreationExpressionSyntax creation => creation.ArgumentList?.Arguments.Select(x => x.GetExpression()) ?? Enumerable.Empty<SyntaxNode>(),
+            AttributeSyntax attr => attr.ArgumentList?.Arguments.Select(x => x.GetExpression()),
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs]: Did a roll back on these.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs:3]: This should just work now:
```suggestion
using EFCC = System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverageAttribute;

...

[EFCC] // Noncompliant
[EFCC(Justification = "justification")] // Compliant
```
Please add the two cases.

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class ExcludeFromCodeCoverageAttributesNeedJustificationBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S1123"; // TODO: Update DiagnosticID.
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:26]:
```suggestion
    private const string DiagnosticId = "S6513";
```

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:26]: Please have a look at the RSpec as well:
https://github.com/SonarSource/rspec/pull/1624

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (NoJustification(c.Node, c.SemanticModel)
+                    && c.SemanticModel.GetSymbolInfo(c.Node).Symbol is IMethodSymbol attribute
+                    && attribute.IsInType(KnownType.System_Diagnostics_CodeAnalysis_ExcludeFromCodeCoverageAttribute)
+                    && HasJustificationProperty(attribute.ContainingType))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:43]:
```suggestion
                    && HasJustificationProperty(attribute.ContainingType)) // "Justification" was added in .Net 5 while the attribute itself is in netstandard2.0
```

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                    c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                }
+            },
+            Language.SyntaxKind.Attribute);
+
+    private bool NoJustification(SyntaxNode node, SemanticModel model) =>
+        GetJustificationExpression(node) is not { } justification
+        || string.IsNullOrWhiteSpace(Language.FindConstantValue(model, justification) as string);
+
+    private static bool HasJustificationProperty(INamedTypeSymbol symbol) => symbol.MemberNames.Contains("Justification");
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:54]:
```suggestion
    private static bool HasJustificationProperty(INamedTypeSymbol symbol) =>
        symbol.MemberNames.Contains(JustificationPropertyName);
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Looks good. Some polishing and edge cases only.

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 8]
```diff
@@ -47,9 +47,11 @@ public abstract class ExcludeFromCodeCoverageAttributesNeedJustificationBase<TSy
             },
             Language.SyntaxKind.Attribute);
 
+    /// <summary>"Justification" was added in .Net 5 whileExcludeFromCodeCoverage in netstandard2.0.</summary>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:50]: This comment should go to the `HasJustificationProperty` method.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Comment needs to be moved.

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class ExcludeFromCodeCoverageAttributesNeedJustificationBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6513";
+
+    /// <summary>"Justification" was added in .Net 5, while ExcludeFromCodeCoverage in netstandard2.0.</summary>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:28]:
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                    c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                }
+            },
+            Language.SyntaxKind.Attribute);
+
+    private bool NoJustification(SyntaxNode node, SemanticModel model) =>
+        GetJustificationExpression(node) is not { } justification
+        || string.IsNullOrWhiteSpace(Language.FindConstantValue(model, justification) as string);
+
+    private static bool HasJustificationProperty(INamedTypeSymbol symbol) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ExcludeFromCodeCoverageAttributesNeedJustificationBase.cs:56]:
```suggestion
    /// <summary>"Justification" was added in .Net 5, while ExcludeFromCodeCoverage in netstandard2.0.</summary>
    private static bool HasJustificationProperty(INamedTypeSymbol symbol) =>
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
The comment should be on HasJustificationProperty

**Comment by martin-strecker-sonarsource** [comment]
Hi @Corniel,

to finish the PR, We will need to import the RSpec into this PR. Can you run the Rspec script like so:
`./scripts/rspec/rspec -language vbnet -ruleKey S6513 -rspecBranch "rule/add-RSPEC-S6513"`
`./scripts/rspec/rspec -language cs -ruleKey S6513 -rspecBranch "rule/add-RSPEC-S6513"`

See also the [readme](https://github.com/SonarSource/sonar-dotnet/tree/master/scripts/rspec#how-to-use) in the spec directory for details.

Please let me know if you have trouble running the script. I can do this for you too.

**Comment by Corniel** [comment]
> ./scripts/rspec/rspec -language vbnet -ruleKey S6513 -rspecBranch "rule/add-RSPEC-S6513"

I have no access to `https://repox.jfrog.io/repox/sonarsource-private-releases/com/sonarsource/rule-api/rule-api/`. I guess this process is not designed to be done by external contributors. ;)

**Comment by martin-strecker-sonarsource** [comment]
@Corniel thanks for trying. I've run the script and pushed the changes (don't forget to pull the changes). I will run the pipeline now.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM
